### DISCUSSION TITLE
Module - M ERROR: Input file is not a msh file!

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/neper#285] [Q&A] [answered]
Hi, I tried to run the FEPX tutorial examples but kept getting this error:  Input file is not a msh file! What's strange is that this error message occurs at random stages, i.e., sometimes at 21% of 2D meshing, sometimes at 53% etc. The same thing also happens when I tried to reproduce examples successfully run by other people. In fact, I have so far only successfully meshed a 10-grains-cube with hex elements and when switching to tet elements it still fails. For these runs, sometimes the message is "Input file is not a msh file!", sometimes it's "Meshing of face #(a number) failed". I expect the latter error message is normal but the strange thing, again, is that other people can successfully run the same code.

I have no clue why this is happening and I hope you guys can give me some insights. I have searched this discussion board but no luck. I'm attaching a sample output below. I'm running the codes on mac Catalina terminal. Both Nerper and Gmsh version at the most recent, 4.2.1 and 4.8.4, respectively.

```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.2.1-7
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 4 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2021, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/Users/Xiao/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] simulation.tess -order 2 -part 4
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `simulation.tess'...
Info   :     [i] Parsed file `simulation.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.1357) 100%
Info   :   - 0D meshing... 100%
Info   :   - 1D meshing... 100%
Info   :   - 2D meshing...  21% (0.72|0.84/79|16| 5)Error  : Input file is not a msh file!
Abort trap: 6
```

TIA!

**Comment by CsatiZoltan** [comment]
@xshang93 If you give me a failing example, I can try running it on my configuration. Maybe we can find the cause.

**Reply by xshang93** [reply]
Thanks for offering to run it but I'm pretty sure the code is right, as they are just the examples provided by FEPX. (https://github.com/acmelab-ua/FEPX/tree/main/examples).
 I did some more research in this discussion board and:
1. The "Input file is not a msh file" might be due to my gmsh installation. I was using homebrew installed gmsh. After I reinstalled it with pip, it seemed to solve this but I still got "Meshing of face #(a number) failed". However, I'm not sure exactly why this worked and it would be nice if someone could give some insights.
2. Someone mentioned that there is an known issue with multithreading on Windows. I'm running my code on mac but I thought I'd give it a try and it actually worked. When I set the thread to 1, the code ran correctly. Anything other than 1 led to the same error message. However, without multithreading it's very slow and I'm not sure if there is a way to improve this?

**Reply by CsatiZoltan** [reply]
1. I only use Neper for post-processing (visualization and fetching raw results from FEPX) only, so I don't know how the tessellation or the meshing works.
1. Unfortunately, I cannot help in that either. I only have Ubuntu installed.

**Comment by rquey** [comment, marked as answer]
Definitely a multithreading issue, as you have already found out. See https://github.com/rquey/neper/discussions/272#discussioncomment-1037530 for a possible solution.
